src/channel/mpmc.rs

```diff
@@ -363,51 +379,120 @@ where A: RingBuf<Item=T>
         }
     }
 
+    /// Returns a stream that will receive values from this channel.
+    ///
+    /// If the channel gets closed, the stream will terminate.
+    pub fn stream(&self) -> ChannelStream<MutexType, T, A> {
+        ChannelStream {
+            channel: Some(self),
+            future: None,
+        }
+    }
+
     /// Closes the channel.
     /// All pending and future send attempts will fail.
     /// Receive attempts will continue to succeed as long as there are items
     /// stored inside the channel. Further attempts will fail.
     pub fn close(&self) {
         self.inner.lock().close()
     }
+
+    /// Returns whether the channel is closed.
+    pub fn is_closed(&self) -> bool {
+        self.inner.lock().is_closed()
+    }
 }
 
-impl <MutexType: RawMutex, T, A> ChannelSendAccess<T> for GenericChannel<MutexType, T, A>
-where A: RingBuf<Item=T>
+impl<MutexType: RawMutex, T, A> ChannelSendAccess<T> for GenericChannel<MutexType, T, A>
+where
+    A: RingBuf<Item = T>,
 {
     unsafe fn try_send(
         &self,
         wait_node: &mut ListNode<SendWaitQueueEntry<T>>,
         cx: &mut Context<'_>,
-    ) -> (Poll<()>, Option<T>)
-    {
+    ) -> (Poll<()>, Option<T>) {
         self.inner.lock().try_send(wait_node, cx)
     }
 
-    fn remove_send_waiter(&self, wait_node: &mut ListNode<SendWaitQueueEntry<T>>)
-    {
+    fn remove_send_waiter(&self, wait_node: &mut ListNode<SendWaitQueueEntry<T>>) {
         self.inner.lock().remove_send_waiter(wait_node)
     }
 }
 
-impl <MutexType: RawMutex, T, A> ChannelReceiveAccess<T> for GenericChannel<MutexType, T, A>
-where A: RingBuf<Item=T>
+impl<MutexType: RawMutex, T, A> ChannelReceiveAccess<T> for GenericChannel<MutexType, T, A>
+where
+    A: RingBuf<Item = T>,
 {
     unsafe fn try_receive(
         &self,
         wait_node: &mut ListNode<RecvWaitQueueEntry>,
         cx: &mut Context<'_>,
-    ) -> Poll<Option<T>>
-    {
+    ) -> Poll<Option<T>> {
         self.inner.lock().try_receive(wait_node, cx)
     }
 
-    fn remove_receive_waiter(&self, wait_node: &mut ListNode<RecvWaitQueueEntry>)
-    {
+    fn remove_receive_waiter(&self, wait_node: &mut ListNode<RecvWaitQueueEntry>) {
         self.inner.lock().remove_receive_waiter(wait_node)
     }
 }
 
+/// A stream that receives from a `GenericChannel`.
+#[derive(Debug)]
+pub struct ChannelStream<'a, MutexType: RawMutex, T, A>
+where
+    A: RingBuf<Item = T>,
+{
+    channel: Option<&'a GenericChannel<MutexType, T, A>>,
+    future: Option<ChannelReceiveFuture<'a, MutexType, T>>,
```

Comment:
That definitely seems the way to go for a `Future` -> `Stream` adapter. From a pinning point of view it looks correct, since the `Stream` also needs to be pinned.

What I'm however not yet sure is whether it's ok from a logical point of view. If a user would start polling on the `Stream` but not getting an item, the `Future` would still be stuck in a signaled state in the Stream. If that prevents other waiters on the `Stream` from getting woken up it might be problematic. I think I would need to review the acutal wakeup logic to understand the impact - e.g. whether the `Channel` kept track of `Future`s it notified but which did not yet pick up an item. The `Mutex` implementation definitely does this in order to guarantee fairness. The  `Channel` might need for rendevousz semantics? 

It had been a while since I last looked into it. Will do on the weekend.

Reply:
So a scenario would be user create a stream, polls the stream once, which returns `Poll::NotReady` then drops the stream future. The inner future is stuck in the stream and might be blocking subsequent futures.

Reply:
Right. This can also happen with the `Future` based API. But there you are a lot more likely to see that an old Future is still around - especially since you can't move the Futures. In the Streams it is somewhat hidden. 

But now since I wrote that: At least the Stream also needs to be pinned and can not be moved - so the user moving around the Stream without knowing that some registration is still active is not possible.